src/backend/executor/nodeHashjoin.c

```diff
@@ -2153,3 +2170,230 @@ ExecHashJoinInitializeWorker(HashJoinState *state,
+	foreach (lc, hj->hashclauses)
+	{
+		Expr *expr = (Expr *)lfirst(lc);
+
+		if (!IsEqualOp(expr))
+			continue;
+
+		lattno = -1;
+		rattno = -1;
+		if (!CheckEqualArgs(expr, &lattno, &rattno))
+			continue;
+
+		if (lattno < 1 || rattno < 1)
+			continue;
+
+		PlanState *target = FindTargetAttr(hjstate, lattno, &lattno);
+		if (lattno == -1 || target == NULL || IsA(target, HashJoinState))
+			continue;
+		Assert(IsA(target, SeqScanState));
+
+		AttrFilter *af = CreateAttrFilter(target, lattno, rattno,
+										  hstate->ps.plan->plan_rows);
+		if (af->bf)
+			hstate->filters = lappend(hstate->filters, af);
+		else
+			pfree(af);
+	}
+}
+
+static bool
+IsEqualOp(Expr *expr)
+{
+	if (!IsA(expr, OpExpr) && !IsA(expr, FuncExpr))
+		return false;
+
+	Oid funcid = InvalidOid;
+	if (IsA(expr, OpExpr))
+		funcid = ((OpExpr *)expr)->opfuncid;
+	if (IsA(expr, FuncExpr))
+		funcid = ((FuncExpr *)expr)->funcid;
+
+	if (funcid == F_INT2EQ  || funcid == F_INT4EQ  || funcid == F_INT8EQ
+		|| funcid == F_INT24EQ || funcid == F_INT42EQ
+		|| funcid == F_INT28EQ || funcid == F_INT82EQ
+		|| funcid == F_INT48EQ || funcid == F_INT84EQ
+	   )
+		return true;
+	else
+		return false;
+}
+
+/*
+ * runtime filters which can be pushed down:
+ * 1. hash expr MUST BE equal op;
+ * 2. args MUST BE Var node;
+ * 3. the data type MUST BE integer;
+ */
+static bool
+CheckEqualArgs(Expr *expr, AttrNumber *lattno, AttrNumber *rattno)
+{
+	if (lattno == NULL || rattno == NULL)
+		return false;
+
+	if (!IsA(expr, OpExpr) && !IsA(expr, FuncExpr))
+		return false;
+
+	List *args = NIL;
+	if (IsA(expr, OpExpr))
+		args = ((OpExpr *)expr)->args;
+	if (IsA(expr, FuncExpr))
+		args = ((FuncExpr *)expr)->args;
+
+	if (!args || list_length(args) != 2)
+		return false;
+
+	ListCell *lc;
+	bool match = false;
+	foreach (lc, args)
+	{
+		match = false;
+
+		if (!IsA(lfirst(lc), Var))
+			break;
+
+		Var *var = lfirst(lc);
+		if (var->varno == INNER_VAR)
+		{
+			*rattno = var->varattno;
+		}
+		else if (var->varno == OUTER_VAR)
+		{
+			*lattno = var->varattno;
+		}
+		else
+			break;
+
+		match = true;
+	}
+
+	return match;
+}
+
+/*
+ * it's just allowed like this:
+ *   HashJoin
+ *      ... a series of HashJoin nodes
+ *        HashJoin
+ *          SeqScan <- target
+ */
+static PlanState *
+FindTargetAttr(HashJoinState *hjstate, AttrNumber attno, AttrNumber *lattno)
+{
+	TargetEntry *te;
+
+	PlanState *parent = (PlanState *)hjstate;
+	PlanState *child  = outerPlanState(hjstate);
+	Assert(child);
+
+	while (child)
+	{
+		/* target is seqscan */
+		if (IsA(child, SeqScanState))
+		{
+			te = (TargetEntry *)list_nth(child->plan->targetlist, attno - 1);
+			if (!IsA(te->expr, Var))
+			{
+				*lattno = -1;
+				return NULL;
+			}
+			else
+			{
+				Var *var = castNode(Var, te->expr);
+
+				/* system column is not allowed */
+				if (var->varattno <= 0)
+				{
+					*lattno = -1;
+					return NULL;
+				}
+
+				*lattno = var->varattno;
+				return child;
```

Comment:
Please remove the `else` and make the code less indented.

Reply:
fix it.